sound/soc/sof/sof-audio.c

```diff
@@ -167,6 +167,296 @@ static int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
+				ret = -ENOMEM;
+				goto widget_free;
+			}
+
+			list_add(&temp_pipe_widget->list, &spcm->stream[dir].pipeline_list);
+		}
+	}
+
+	/*
+	 * error in setting pipeline connections will result in route status being reset for
+	 * routes that were successfully set up when the widgets are freed.
+	 */
+	ret = sof_setup_pipeline_connections(sdev, list, dir);
+	if (ret < 0)
+		goto widget_free;
+
+	/* setup pipeline widgets and complete pipelines */
+	list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+		struct snd_sof_widget *pipe_widget = sof_find_widget_by_comp_id(sdev, w->comp_id);
+
+		ret = sof_widget_setup(sdev, pipe_widget);
+		if (ret < 0)
+			goto pipe_free;
+		else
+			pipe_widget_count++;
+
+		if (pipe_widget->complete)
+			continue;
+
+		pipe_widget->complete = snd_sof_complete_pipeline(sdev->dev, pipe_widget);
+		if (pipe_widget->complete < 0) {
+			ret = pipe_widget->complete;
+			goto pipe_free;
+		}
+	}
+
+	return 0;
+
+pipe_free:
+	/* free the pipe widgets that were setup successfully set up */
+	list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+		struct snd_sof_widget *pipe_widget = sof_find_widget_by_comp_id(sdev, w->comp_id);
+
+		if (j == pipe_widget_count)
+			break;
+
+		if (pipe_widget)
+			sof_widget_free(sdev, pipe_widget);
+		j++;
+	}
+
+widget_free:
+	/* free all widgets that have been setup successfully */
+	for_each_dapm_widgets(list, j, widget) {
+		/* i is the count of the widgets that were setup successfully */
+		if (j == i)
+			break;
+
+		if (is_sof_widget(widget))
+			sof_widget_free(sdev, widget->dobj.private);
+	}
+
+	/* free the temp_widgets in the pipeline list */
+	list_for_each_entry_safe(w, _w, &spcm->stream[dir].pipeline_list, list) {
+		list_del(&w->list);
+		kfree(w);
+	}
+
+	return ret;
+}
+
+struct snd_sof_widget *sof_find_widget_by_comp_id(struct snd_sof_dev *sdev, int comp_id)
+{
+	struct snd_sof_widget *swidget;
+
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (swidget->comp_id == comp_id)
+			return swidget;
+
+	return NULL;
+}
+
+int sof_widget_list_free(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget_list *list,
+			 struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *w, *_w;
+	int i;
+
+	/*
+	 * Free widgets in the list. This can fail but continue freeing other widgets to keep
+	 * use_counts balanced.
+	 */
+	for_each_dapm_widgets(list, i, widget)
+		if (is_sof_widget(widget))
+			sof_widget_free(sdev, widget->dobj.private);
+
+	/* free pipeline widgets */
+	list_for_each_entry_safe(w, _w, &spcm->stream[dir].pipeline_list, list) {
+		struct snd_sof_widget *pipe_widget = sof_find_widget_by_comp_id(sdev, w->comp_id);
+
+		/* this can fail but continue freeing other widgets to keep use_counts balanced */
+		if (pipe_widget)
+			sof_widget_free(sdev, pipe_widget);
+
+		list_del(&w->list);
+		kfree(w);
+	}
+
+	snd_soc_dapm_dai_free_widgets(&list);
+	spcm->stream[dir].list = NULL;
```

Comment:
this seems somewhat inconsistent to me. This function is always called with the second parameter `list` obtainable from the last two parameters as `spcm->stream[dir].list`. Would it make sense to ever call this function differently? If not, I'd drop the second parameter, that would also make this `NULL` assignment more logical. Similarly then you'd change the `sof_widget_list_setup()` function slightly - see next patch. Or make the second parameter to `&spcm->stream[dir].list` and make this assignment `*list = NULL;`

Reply:
ack, i removed the list parameter from both functions now.